bentoml/cli/bento.py

```diff
@@ -237,3 +245,112 @@ def retrieve(bento, target_dir):
+        You can also optionally provide a `--push` flag, which will push the built
+        image to the Docker repository specified by the `--docker-repository`.
+
+        If you would like to push to Docker Hub, `--docker-repository` can just be
+        your Docker Hub username. Otherwise, the tag should include the hostname as
+        well. For example, for Google Container Registry, the `--docker-repository`
+        would be `[HOSTNAME]/[PROJECT-ID]`
+        """
+        name, version = bento.split(':')
+        yatai_client = YataiClient()
+
+        get_bento_result = yatai_client.repository.get(name, version)
+        if get_bento_result.status.status_code != status_pb2.Status.OK:
+            error_code, error_message = status_pb_to_error_code_and_message(
+                get_bento_result.status
+            )
+            raise CLIException(
+                f'BentoService {name}:{version} not found - '
+                f'{error_code}:{error_message}',
+            )
+
+        if get_bento_result.bento.uri.s3_presigned_url:
+            bento_service_bundle_path = get_bento_result.bento.uri.s3_presigned_url
+        else:
+            bento_service_bundle_path = get_bento_result.bento.uri.uri
+
+        _echo(f"Found Bento: {bento_service_bundle_path}")
+
+        if docker_repository is not None:
+            name = f'{docker_repository}/{name}'
+
+        name, version = make_bento_name_docker_compatible(name, version)
+        tag = f"{name}:{version}"
+        if tag != bento:
+            _echo(
+                f'Bento name or tag was changed to be Docker compatible. \n'
+                f'"{bento}" -> "{tag}"',
+                CLI_COLOR_WARNING,
+            )
+
+        docker_api = docker.APIClient()
+        try:
+            with Spinner(f"Building Docker image: {name}\n"):
+                _echo_docker_api_result(
+                    docker_api.build(
+                        path=bento_service_bundle_path, tag=tag, decode=True,
+                    )
+                )
+        except docker.errors.APIError as error:
+            raise CLIException(f'Could not build Docker image: {error}')
+
+        _echo(
+            f'Finished building {tag} from {bento}', CLI_COLOR_SUCCESS,
+        )
+
+        if push:
+            if not docker_repository:
+                raise CLIException('Docker Registry must be specified when pushing.')
+
+            auth_config_payload = (
+                {"username": username, "password": password}
+                if username or password
+                else None
+            )
+
+            try:
+                with Spinner(f"Pushing docker image to {tag}\n"):
+                    _echo_docker_api_result(
+                        docker_api.push(
+                            repository=name,
+                            tag=version,
+                            stream=True,
+                            decode=True,
+                            auth_config=auth_config_payload,
```

Comment:
when username and password are not provided via the CLI, does `docker_api.push` pick up the local docker authentication information?

Reply:
yes